subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java

```diff
@@ -168,6 +172,22 @@ public Set<Object> getFrom() {
         return filesWrapper;
     }
 
+    @Override
+    public void setFromAnyValue(Object object) {
+        // TODO: Move check in setFrom
+        // Don't allow a += b or a = (a + b), this is not support and results in stack overflow
+        if (object instanceof UnionFileCollection && ((UnionFileCollection) object).getSources().contains(this)) {
+            throw new UnsupportedOperationException("Operation for ConfigurableFileCollection of type a += b or a = (a + b) is not supported");
+        }
```

Comment:
:x: Let's not rely on `UnionFileCollection` specifically, and visit the `FileCollection` structure instead to see if of the descendants are `this`.

Reply:
Used `fileCollection.visitStructure`, please review that carefuly if that is a correct approach

Reply:
Looks good to me. The code is now more straightforward about what is happening, though it's also much more verbose due to the no-op method implementations. I guess that's fine.